kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClientHandler.java

```diff
@@ -47,15 +60,28 @@ public void channelReadComplete(ChannelHandlerContext ctx) {
 
     @Override
     public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        cause.printStackTrace();
         ctx.close();
     }
 
-    /**
-     * A future that is completed when the response is received
-     * @return on response future
-     */
-    public CompletableFuture<DecodedResponseFrame<?>> getOnResponseFuture() {
-        return onResponse;
+    public CompletableFuture<DecodedResponseFrame<?>> sendRequest(DecodedRequestFrame<?> decodedRequestFrame) {
+        queue.addLast(decodedRequestFrame);
+        processPendingWrites();
+        return decodedRequestFrame.getResponseFuture();
+    }
+
+    private void processPendingWrites() {
+        ctx.executor().execute(() -> {
+            if (ctx.channel().isActive()) {
+                while (queue.peek() != null) {
+                    DecodedRequestFrame<?> msg = queue.removeFirst();
+                    ctx.writeAndFlush(msg).addListener(c -> {
+                        if (msg.body() instanceof ProduceRequest && ((ProduceRequest) msg.body()).acks() == 0) {
+                            msg.getResponseFuture().complete(null);
+                        }
+                    });
+                }
+            }
+        });
```

Comment:
I don't have a good picture of when the channel would be active or inactive but it feels strange to gate entering the loop on if the channel is active or not. I would have thought we would do that inside the loop.
```suggestion
            while (queue.peek() != null) {
                if (ctx.channel().isActive()) {
                    DecodedRequestFrame<?> msg = queue.removeFirst();
                    ctx.writeAndFlush(msg).addListener(c -> {
                        if (msg.body() instanceof ProduceRequest && ((ProduceRequest) msg.body()).acks() == 0) {
                            msg.getResponseFuture().complete(null);
                        }
                    });
                }
            }
        });
```

Reply:
^ would be an infinite loop of peeking and checking active if the channel died. But yeah maybe it should drain the queue and if the channel is closed complete the future exceptionally?

Reply:
Hmm I certainly think we should be explicit about what happens if the connection is closed. 

if the connection is closed during the draining of the queue we would end up with logs of exceptions I think so maybe need to test the channel is active and break if not during the loop as well.

Reply:
I've refactored to propagate the failed write exception to the response future.